drivers: timer: nrf_rtc_timer: user channels count dependency inv

ankuns opened the pull request:
This PR is related to the https://github.com/zephyrproject-rtos/zephyr/issues/67117
and is inspired by https://github.com/zephyrproject-rtos/zephyr/pull/65908 .

For calculation of number of user channels of nrf_rtc_timer a new mechanism allowing dependency inversion is introduced.

The `CONFIG_NRF_RTC_TIMER_USER_CHAN_COUNT` is no longer used directly in the C code. Instead it is replaced by the
`K_NRF_RTC_TIMER_USER_CHAN_COUNT` macro which is calculated automatically during the build.

Users can add own requirements on the `K_NRF_RTC_TIMER_USER_CHAN_COUNT` macro by:
- setting the value of the Kconfig option `NRF_RTC_TIMER_USER_CHAN_COUNT` (legacy appoach).
- providing Kconfig options named according to pattern `NRF_RTC_TIMER_USER_CHAN_COUNT_ADD_<custom_name>`, each of which increases the value of `K_NRF_RTC_TIMER_USER_CHAN_COUNT` by given value.
- providing Kconfig options named according to pattern `NRF_RTC_TIMER_USER_CHAN_COUNT_MIN_<custom_name>`, each of which imposes limit on the minimum value of the `K_NRF_RTC_TIMER_USER_CHAN_COUNT` macro.

Mentioned new possible Kconfig options can be added in multiple modules, also outside of Zephyr tree or in downstream repositories. They express distributed requirements on value of `K_NRF_RTC_TIMER_USER_CHAN_COUNT`.

The definition of the Kconfig option `NRF_RTC_TIMER_USER_CHAN_COUNT` does not need to know now every possible user (including known downstream repositories) to set up reasonable default value. It is the nrf_rtc_timer module users' responsibility to express requirements in their own modules. Because of that default values 2 and 3 depending on NRF_802154_RADIO_DRIVER have been removed. In the pure Zephyr at the current state the nRF 802.15.4 does not require any user channel of nrf_rtc_timer. If user channels are required by projects based on Zephyr (including more sophisticated implementation of platform support code for nRF 802.15.4 Radio Driver), individual requirements can be easily expressed in separate repositories.

ankuns committed:
drivers: timer: nrf_rtc_timer: user channels count dependency inv

For calculation of the number of user channels of nrf_rtc_timer a new
mechanism allowing dependency inversion is used.

The `CONFIG_NRF_RTC_TIMER_USER_CHAN_COUNT` is no longer used directly
in the C code. Instead it is replaced by the
`K_NRF_RTC_TIMER_USER_CHAN_COUNT` macro which is calculated
automatically during the build by the
`zephyr_compile_definition_kconfig_based` cmake function.

Users can add own requirements on the `K_NRF_RTC_TIMER_USER_CHAN_COUNT`
macro by:
- setting the value of the Kconfig option
  `NRF_RTC_TIMER_USER_CHAN_COUNT` (legacy appoach).
- providing Kconfig options named according to pattern
  `NRF_RTC_TIMER_USER_CHAN_COUNT_ADD_*`, each of which
  increases the value of `K_NRF_RTC_TIMER_USER_CHAN_COUNT` by given
  value.
- providing Kconfig options named according to pattern
  `NRF_RTC_TIMER_USER_CHAN_COUNT_MIN_*`, each of which
  imposes limit on the minimum value of the
  `K_NRF_RTC_TIMER_USER_CHAN_COUNT` macro.

Mentioned new possible Kconfig options can be added in multiple modules,
also outside of Zephyr tree or in downstream repositories. They express
distributed requirements on value of `K_NRF_RTC_TIMER_USER_CHAN_COUNT`.

The definition of the Kconfig option `NRF_RTC_TIMER_USER_CHAN_COUNT`
does not need to know now every possible user (including known
downstream repositories) to set up reasonable default value.
It is the nrf_rtc_timer module users' responsibility to express
requirements in their own modules. Because of that default values
2 and 3 depending on NRF_802154_RADIO_DRIVER have been removed.
In the pure Zephyr at the current state the nRF 802.15.4 does not
require any user channel of nrf_rtc_timer. If user channels
are required by projects based on Zephyr (including more sophisticated
implementation of platform support code for nRF 802.15.4 Radio Driver),
individual requirements can be easily expressed in separate
repositories.

Signed-off-by: Andrzej Kuroś <andrzej.kuros@nordicsemi.no>

jciupis commented:
I think the implementation follows your concept nicely. I do have two comments about the concept itself though.

You decided to provide two separate mechanisms to for expressing requirements on the value of `K_NRF_RTC_TIMER_USER_CHAN_COUNT`: one for requesting/reserving a specified number of a resource instances, and the other for specifying a total minimum for the number of all requested instances. The intention behind the latter is unclear to me. It seems that such a pattern could be used to allow a single resource instance to be reused instead of reserving a separate instance. However, there's no policy on _how_ such reuse is expected to work. As a result, any module that would like to reuse a single resource instance and express it using the syntax you provided would need to know and understand other clients in the system that it might want to reuse the instance with to determine if that's at all possible.

Let's consider an example from the current implementation of the PR where you used the pattern in question to specify the minimum required number of additional channels for sync_rtc module as 1. This module only needs the channel at initialization, which allows the channel (i.e. the resource) to be reused later, when the main program function starts. However, what would happen if there was a different module, let's temporarily name it B, that uses the resource in the same way? B's developer would follow the same reasoning as you did in the PR, and they would only specify the minimum required number of additional channels as 1, so that the resource can be reused by other modules later. As a result, there would be an implicit coupling between sync_rtc and B, which could cause any one of them to be unable to acquire the resource despite correctly specifying their requirements. That could happen always, never, or non-deterministically, depending on the details of their implementations.

The purpose of this PR is to remove coupling between clients of the same resource and it seems to me that this pattern breaks this promise.

The second comment I have regarding the proposed concept is that I don't see any mechanism for controlling the upper bound of the resource instances that can be requested. In case of finite resources such as peripheral channels/instances etc. such an upper bound should be possible to specify at build-time. In my opinion, if it's possible to detect a wrong configuration at build time and throw a build error, that's preferable over delegating the responsibility for detecting and acting on such a wrong configuration to runtime.

I understand that the current implementation also doesn't provide such a mechanism. However, currently the application developer is responsible for setting it to a correct value. With the proposed approach, it happens behind the scenes using values distributed throughout the entire system, which makes it more difficult to investigate potential problems.

Let me know what you think on these two potential issues.

ankuns commented:
@jciupis to your https://github.com/zephyrproject-rtos/zephyr/pull/67854#issuecomment-1903664189:

**About 1st issue:** 
generally agree that there is the risk you mentioned, but it existed before already, and I don't attempt to fix it.
Please have a look here (code before my change):
https://github.com/zephyrproject-rtos/zephyr/blob/db429004fdea646289bdb778e40a171cacfd16cf/soc/arm/nordic_nrf/nrf53/Kconfig.sync_rtc#L23-L26
If CONFIG_NRF_802154_RADIO_DRIVER=y the nrf_rtc_channel was still "borrowed" from the channels pool required by the NRF_802154_RADIO_DRIVER. The difference is that sync_rtc uses the rtc channel during the system initialization and the nrf_802154 uses the rtc channels later.

I'm open for suggestions about fixing this issue.

**About 2nd issue:** there is mechanism for nrf_rtc_timer to check upper bound, it is BUILD_ASSERT here:
https://github.com/ankuns/zephyr/blob/9709255774d08a4e862b8792d788279e6246ef56/drivers/timer/nrf_rtc_timer.c#L31

ankuns commented:
@nordic-krch @anangl @tejlmand What do you think about proposed approach? (PR is a draft because I don't want it to be merged yet, I wanted to discuss it with you first).

The commit is cherry-picked to downstream sdk-zephyr to see how much does it simplify, you can follow it.

I know I could make the code I put into drivers/timer/CMakeLists.txt into a cmake function which would be called like:
```cmake 
zephyr_compile_definition_kconfig_based(
    K_NRF_RTC_TIMER_USER_CHAN_COUNT  /* macro that is being provided to zephyr, calation based on Kconfigs */
        KCONFIG_INIT CONFIG_NRF_RTC_TIMER_USER_CHAN_COUNT      /* Kconfig being initial value of the macro */
        KCONFIG_ADD CONFIG_NRF_RTC_TIMER_USER_CHAN_COUNT_ADD_  /* optional Kconfig prefix for expressing increase of the macro value */
        KCONFIG_MIN CONFIG_NRF_RTC_TIMER_USER_CHAN_COUNT_MIN_  /* optional Kconfig prefixfor expressing minimal value of the macro */
)
```
So that the pattern could be easily re-used.
I will do that if you generally agree for such change.The mentioned cmake function could be later be used to express in a similar way requirements on stack sizes or "sizes"/"capacities" of other resources that could be fine-trimmed build-time, but it's hard do do it properly taking into account dependencies.

ankuns committed:
cmake: extensions: introduce zephyr_compile_definition_kconfig_based

The cmake function `zephyr_compile_definition_kconfig_based`
is provided to allow automatic calculation of macros based on
scattered Kconfig options.

Signed-off-by: Andrzej Kuroś <andrzej.kuros@nordicsemi.no>
